File: app/src/test/java/com/example/myapplication/core/program/ProgramPhasePlannerTest.kt
package com.example.myapplication.core.program

import com.example.myapplication.core.program.ProgramPhase.BUILD
import com.example.myapplication.core.program.ProgramPhase.CONSOLIDATE
import com.example.myapplication.core.program.ProgramPhase.DELOAD
import com.example.myapplication.core.program.ProgramPhase.FOUNDATION
import org.junit.Assert.assertEquals
import org.junit.Assert.assertThrows
import org.junit.Test

class ProgramPhasePlannerTest {
    @Test
    fun `eight week plan uses foundation build consolidate and deload phases`() {
        val phases = (1..8).map { ProgramPhasePlanner.phaseFor(it, 8) }

        assertEquals(
            listOf(FOUNDATION, FOUNDATION, BUILD, BUILD, BUILD, CONSOLIDATE, CONSOLIDATE, DELOAD),
            phases,
        )
    }

    @Test
    fun `short plan does not force a deload week`() {
        assertEquals(listOf(CONSOLIDATE), (1..1).map { ProgramPhasePlanner.phaseFor(it, 1) })
        assertEquals(
            listOf(BUILD, BUILD, CONSOLIDATE),
            (1..3).map { ProgramPhasePlanner.phaseFor(it, 3) },
        )
    }

    @Test
    fun `phase rejects a week outside program duration`() {
        assertThrows(IllegalArgumentException::class.java) { ProgramPhasePlanner.phaseFor(0, 8) }
        assertThrows(IllegalArgumentException::class.java) { ProgramPhasePlanner.phaseFor(9, 8) }
    }
}

File: flutter/test/core/program/program_phase_planner_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:gym_app/core/program/program_phase_planner.dart';

void main() {
  group('ProgramPhasePlannerTest', () {
    test('eight week plan uses foundation build consolidate and deload phases',
        () {
      final phases =
          List.generate(8, (i) => ProgramPhasePlanner.phaseFor(i + 1, 8));

      expect(
        phases,
        equals(const [
          ProgramPhase.foundation,
          ProgramPhase.foundation,
          ProgramPhase.build,
          ProgramPhase.build,
          ProgramPhase.build,
          ProgramPhase.consolidate,
          ProgramPhase.consolidate,
          ProgramPhase.deload,
        ]),
      );
    });

    test('short plan does not force a deload week', () {
      expect(
        List.generate(1, (i) => ProgramPhasePlanner.phaseFor(i + 1, 1)),
        equals(const [ProgramPhase.consolidate]),
      );
      expect(
        List.generate(3, (i) => ProgramPhasePlanner.phaseFor(i + 1, 3)),
        equals(const [
          ProgramPhase.build,
          ProgramPhase.build,
          ProgramPhase.consolidate
        ]),
      );
    });

    test('phase rejects a week outside program duration', () {
      expect(() => ProgramPhasePlanner.phaseFor(0, 8), throwsArgumentError);
      expect(() => ProgramPhasePlanner.phaseFor(9, 8), throwsArgumentError);
    });
  });
}
